### DISCUSSION TITLE
Selecting model by where clause with onupdate fields does not get correct result

### DISCUSSION THREAD

**Discussion by kigawas** [sqlalchemy/sqlalchemy#6989] [Usage Questions] [answered]
### Describe the bug

Selecting model by where clause with onupdate fields does not get correct onupdate field values.

Perhaps related with cache strategy.

### To Reproduce

```python
from dataclasses import dataclass
from datetime import datetime
import asyncio

from sqlalchemy import Table, String, Boolean, Column, DateTime, Integer, MetaData, text
from sqlalchemy.future import select
from sqlalchemy.orm import registry
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import clear_mappers, sessionmaker
from sqlalchemy.sql.expression import update

mapper_registry = registry()
metadata = MetaData()


@dataclass
class User:
    id: int  # pk
    name: str
    _created_at: datetime = datetime.now()
    _updated_at: datetime = datetime.now()
    _version: int = 0
    _deleted: bool = False


UserTable = Table(
    "user",
    metadata,
    Column("id", Integer, primary_key=True, autoincrement=True),
    Column("name", String, nullable=False),
    Column(
        "_created_at",
        DateTime,
        nullable=False,
        default=datetime.now,
    ),
    Column(
        "_updated_at",
        DateTime,
        nullable=False,
        default=datetime.now,
        onupdate=datetime.now,
    ),
    Column(
        "_version",
        Integer,
        onupdate=text("_version + 1"),
        nullable=False,
        default=0,
    ),
    Column("_deleted", Boolean, nullable=False, default=False),
)


async def create_db(db_url):
    engine = create_async_engine(db_url)

    async with engine.begin() as conn:
        await conn.run_sync(metadata.drop_all)
        await conn.run_sync(metadata.create_all)

    mapper_registry.map_imperatively(User, UserTable)

    make_session = sessionmaker(
        bind=engine,
        expire_on_commit=False,
        class_=AsyncSession,
    )  # type: ignore

    async with make_session() as session:
        assert isinstance(session, AsyncSession)
        u = User(id=None, name="aaa")  # type:ignore
        session.add(u)
        await session.flush()
        assert u.id == 1

        stmt = (
            update(User)
            .values(dict(name="bbb"))
            .where(User.id == u.id, User._version == 0)
        )
        await session.execute(stmt)
        await session.commit()

        res = await session.execute(select(User).where(User.id == u.id))
        user = res.scalar_one()
        assert user.name == "bbb"
        assert user._version == 1  # AssertionError

    clear_mappers()


async def main():
    await create_db("postgresql+asyncpg://postgres:@localhost/test_temp")


asyncio.run(main())
```

### Error

Actually no error except AssertionError above.

### Versions

- OS: macOS
- Python: 3.9.6
- SQLAlchemy: 1.4.20
- Database: postgresql
- DBAPI: asyncpg

**Comment by CaselIT** [comment, marked as answer]
Hi,

That's the expected behaviour, since you specified `expire_on_commit=False,`. If you add `execution_options={"populate_existing": True},` to execute it will emit sql in any case. You can also use `select.execution_options`. https://docs.sqlalchemy.org/en/14/orm/queryguide.html#orm-queryguide-populate-existing

**Reply by CaselIT** [reply]
@zzzeek I think we miss a list of all execution options accepted in the documentation. Do you think it's something that feasible to add?

**Reply by kigawas** [reply]
Thanks. It looks like `populate_existing` will update all fields, can I specify some columns of the model (like `_version`) as "updated on server side" to make sure they'll always get refreshed? I tried "returning", but it didn't help.

**Reply by CaselIT** [reply]
You can expire only some fields using https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.expire.params.attribute_names.

if you use expire `populate_existing` should not be needed (but I've not tried it)

**Reply by zzzeek** [reply]
Different execution options are part of different contexts such as orm vs core.   I need people to read the orn querying guide for their reference, there should be a list of orm related options there

**Reply by CaselIT** [reply]
I meant maybe a table with something like these headers `option name | orm vs code | description | see also`?

**Reply by zzzeek** [reply]
There's so many docs so adding more would have to be done in some way that it doesn't just add to the chaos

**Reply by CaselIT** [reply]
> 
> 
> There's so many docs so adding more would have to be done in some way that it doesn't just add to the chaos

that's true :)

**Reply by kigawas** [reply]
> Different execution options are part of different contexts such as orm vs core. I need people to read the orn querying guide for their reference, there should be a list of orm related options there

Haven't they been unified as one? For me I mostly use the session.execute in 1.4 version

**Reply by zzzeek** [reply]
core context means you are executing using connection.execute(stmt).  the "populate_existing" option is ignored in this context, as this is a Core context.    Only if you use session.execute(stmt) is "populate_existing" relevant.
